src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Threading;
+using System.Runtime.Versioning;
+
+namespace System.Security.Cryptography
+{
+#if !NET7_0_OR_GREATER && NET
+    [UnsupportedOSPlatform("browser")]
+#endif
+    internal sealed partial class SP800108HmacCounterKdfImplementationManaged : ISP800108HmacCounterKdfImplementation
+    {
+        private byte[] _key;
+        private int _keyReferenceCount = 1;
+        private readonly HashAlgorithmName _hashAlgorithm;
+
+        public void DeriveBytes(ReadOnlySpan<byte> label, ReadOnlySpan<byte> context, Span<byte> destination)
+        {
+            try
+            {
+                byte[] key = IncrementAndAcquireKey();
+                DeriveBytesOneShot(key, _hashAlgorithm, label, context, destination);
+            }
+            finally
+            {
+                ReleaseKey();
+            }
+        }
+
+        public void DeriveBytes(ReadOnlySpan<char> label, ReadOnlySpan<char> context, Span<byte> destination)
+        {
+            try
+            {
+                byte[] key = IncrementAndAcquireKey();
+                DeriveBytesOneShot(key, _hashAlgorithm, label, context, destination);
+            }
+            finally
+            {
+                ReleaseKey();
+            }
+        }
+
+        public void DeriveBytes(byte[] label, byte[] context, Span<byte> destination)
+        {
+            try
+            {
+                byte[] key = IncrementAndAcquireKey();
+                DeriveBytesOneShot(key, _hashAlgorithm, label, context, destination);
+            }
+            finally
+            {
+                ReleaseKey();
+            }
+        }
+
+        public void Dispose()
+        {
+            ReleaseKey();
+        }
+
+        private byte[] IncrementAndAcquireKey()
+        {
+            while (true)
+            {
+                int current = Volatile.Read(ref _keyReferenceCount);
+
+                if (current <= 0)
+                {
+                    throw new ObjectDisposedException(nameof(SP800108HmacCounterKdfImplementationManaged));
+                }
+
+                int incrementedCount = checked(current + 1);
+
+                if (Interlocked.CompareExchange(ref _keyReferenceCount, incrementedCount, current) == current)
+                {
+                    return _key;
+                }
+            }
+        }
+
+        public void ReleaseKey()
+        {
+            while (true)
+            {
+                int current = Volatile.Read(ref _keyReferenceCount);
+
+                if (current <= 0)
+                {
+                    // The key has already been released and zeroed. Do nothing. This may occur when the this
```

Comment:
Technically this is an observable change: calling `Dispose` multiple times could cause the key to be zeroed while an operation is taking place, which would cause the `DeriveBytes` operation to return an _incorrect_ value (as opposed to throwing an exception). A standard `SafeHandle` implementation works around this by special-casing `Dispose` to only decrement the ref count once.

You could work around it with something like this:

```cs
int _disposed = 0;
public void Dispose()
{
    if (Interlocked.Exchange(ref _disposed, 1) == 0) { ReleaseKey(); }
}
```

It also means that you don't need a loop within `ReleaseKey`. The logic becomes a whole lot easier.

```cs
private void ReleaseKey()
{
    int newCount = Interlocked.Decrement(ref _keyReferenceCount);
    Debug.Assert(newCount >= 0, "How did this go negative?");
    if (newCount == 0) { ZeroKey(); }
}
```

`SafeHandle` has more complexity to keep the "was disposed?" bit _within_ the ref count integer, but that seems like overkill here.

Reply:
Ohhhhhhhhhhhh good catch. This uncovered a few problems with the reference counting. This along with some asserts makes me feel better about the whole thing.

Reply:
> but should it be private?

Yes it should have.